Mathlib/CategoryTheory/Category/Cat/CartesianClosed.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2025 Emily Riehl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Emily Riehl
+-/
+import Mathlib.CategoryTheory.Closed.Cartesian
+import Mathlib.CategoryTheory.Functor.Currying
+import Mathlib.CategoryTheory.Monoidal.Cartesian.Cat
+
+/-!
+# Cartesian closed structure on `Cat`
+
+The category of small categories is cartesian closed, with the exponential at a category `C`
+defined by the functor category mapping out of `C`.
+
+Adjoint transposition is defined by currying and uncurrying.
+
+TODO: It would be useful to investigate and formalize further compatibilities along the
+lines of `Cat.ihom_obj` and `Cat.ihom_map`, relating currying of functors with currying in
+monoidal closed categories and precomposition with left whiskering. These may not be
+definitional equalities but may have to be phrased using `eqToIso`.
+
+-/
+
+universe v u v₁ v₂ v₃ v₄ u₁ u₂ u₃ u₄
+
+namespace CategoryTheory
+
+open Functor Cat
+
+namespace Cat
+
+variable (C : Type u) [Category.{v} C]
+
+/-- A category `C` induces a functor from `Cat` to itself defined
+by forming the category of functors out of `C`. -/
+@[simps]
+def exp : Cat ⥤ Cat where
+  obj D := Cat.of (C ⥤ D)
+  map F := (whiskeringRight _ _ _).obj F
+
+end Cat
+
+section
+
+variable {B : Type u₁} [Category.{v₁} B] {C : Type u₂} [Category.{v₂} C] {D : Type u₃}
+  [Category.{v₃} D] {E : Type u₄} [Category.{v₄} E]
+
+/-- The isomorphism of categories of bifunctors given by currying. -/
+@[simps!]
+def curryingIso : Cat.of (C ⥤ D ⥤ E) ≅ Cat.of (C × D ⥤ E) :=
+  isoOfEquiv CategoryTheory.currying Functor.curry_obj_uncurry_obj Functor.uncurry_obj_curry_obj
+
+/-- The isomorphism of categories of bifunctors given by flipping the arguments. -/
+@[simps!]
+def flippingIso : Cat.of (C ⥤ D ⥤ E) ≅ Cat.of (D ⥤ C ⥤ E) :=
+  isoOfEquiv CategoryTheory.flipping Functor.flip_flip Functor.flip_flip
+
+/-- Natural isomorphism witnessing `comp_flip_uncurry_eq`. -/
+@[simps!]
+def compFlipUncurryIso (F : B ⥤ D) (G : D ⥤ C ⥤ E) :
+    uncurry.obj (F ⋙ G).flip ≅ (𝟭 C).prod F ⋙ uncurry.obj G.flip := .refl _
+
+lemma comp_flip_uncurry_eq (F : B ⥤ D) (G : D ⥤ C ⥤ E) :
+    uncurry.obj (F ⋙ G).flip = (𝟭 C).prod F ⋙ uncurry.obj G.flip := rfl
+
+end
+
+section
+variable {B C D E : Type u} [Category.{u} B] [Category.{u} C]
+  [Category.{u} D] [Category.{u} E]
+
+/-- Natural isomorphism witnessing `comp_flip_curry_eq`. -/
+@[simps!]
+def curryObjCompIso (F : C × B ⥤ D) (G : D ⥤ E) :
+    (curry.obj (F ⋙ G)).flip ≅ (curry.obj F).flip ⋙ (Cat.exp (Cat.of C)).map G.toCatHom :=
+  .refl _
+
+lemma curry_obj_comp_flip (F : C × B ⥤ D) (G : D ⥤ E) :
+    (curry.obj (F ⋙ G)).flip =
+      (curry.obj F).flip ⋙ (Cat.exp (Cat.of C)).map G.toCatHom := rfl
```

Comment:
This should be stated using `whiskeringRight` instead of `Cat.exp`, and so it should also be moved to `Functor.Currying`.